### PR TITLE
SONAR-8701 Computation of lines measures should use lines from component report

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/FileAttributes.java [hunk 1]
```diff
@@ -31,10 +33,13 @@
   private final boolean unitTest;
   @CheckForNull
   private final String languageKey;
+  private final int lines;
 
-  public FileAttributes(boolean unitTest, @Nullable String languageKey) {
+  public FileAttributes(boolean unitTest, @Nullable String languageKey, int lines) {
     this.unitTest = unitTest;
     this.languageKey = languageKey;
+    checkArgument(lines > 0, "Lines has not been set for this file");
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/FileAttributes.java]: The text of the error makes assumption of where the constructor is called, which is too much.

See my comment on `getLines()`, you're adding this check to ensure comply with the contract this class advertises, the message should be related to the contract. The error should be "number of line must be >= 0".

If you want to throw this error message (I think it's a good idea), put it where applicable, ie. in ComponentRootBuilder.

BTW, I'm not sure a file can't be empty. See my other comment.

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/FileAttributes.java [hunk 2]
```diff
@@ -46,11 +51,16 @@ public String getLanguageKey() {
     return languageKey;
   }
 
+  public int getLines() {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/FileAttributes.java]: you could document that return value can't be less than 0.

File: server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/component/ComponentImplTest.java [hunk 3]
```diff
@@ -105,15 +105,15 @@ public void getFileAttributes_throws_ISE_if_BatchComponent_does_not_have_type_FI
...
   @Test
   public void isUnitTest_returns_true_if_IsTest_is_set_in_BatchComponent() {
-    ComponentImpl component = buildSimpleComponent(FILE, "file").setFileAttributes(new FileAttributes(true, null)).build();
+    ComponentImpl component = buildSimpleComponent(FILE, "file").setFileAttributes(new FileAttributes(true, null, 1)).build();
 
     assertThat(component.getFileAttributes().isUnitTest()).isTrue();
   }
 
   @Test
   public void isUnitTest_returns_value_of_language_of_BatchComponent() {
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/component/ComponentImplTest.java]: A UT is missing on the IAE raised by constructor when the line parameter is < 0, and another one is missing on the value of the getLine().
That's debt, but there is also no UT for `FileAttributes#toString()` (you modified it, there is related change in UTs, it's like your change does not exist).

File: server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/component/ComponentRootBuilderTest.java [hunk 4]
```diff
@@ -466,18 +470,37 @@ public void createFileAttributes_returns_null_when_type_is_not_FILE() {
...
+    assertThat(createFileAttributes(newBuilder().setType(FILE).setIsTest(true).setLines(1).build()).isUnitTest()).isTrue();
+  }
+
+  @Test
+  public void createFileAttributes_sets_lines_in_Scanner_Component() {
+    assertThat(createFileAttributes(newBuilder().setType(FILE).setLines(10).build()).getLines()).isEqualTo(10);
+  }
+
+  @Test
+  public void fail_with_IAE_when_createFileAttributes_sets_lines_to_0() throws Exception {
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/component/ComponentRootBuilderTest.java]: is it really true? can't a file be empty? I doubt it. please check with scanner.

File: sonar-scanner-protocol/src/main/protobuf/scanner_report.proto [hunk 5]
```diff
@@ -86,7 +86,7 @@ message Component {
   // Only available on PROJECT and MODULE types
   // TODO rename this property -> moduleKey ?
   string key = 10;
-  // Only available on FILE type
+  // Only available on FILE type, should always be at least 1
```
Comment (sns-seb) [inline sonar-scanner-protocol/src/main/protobuf/scanner_report.proto]: make sure of that with scanner. I doubt it.

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SizeMeasuresStep.java [hunk 6]
```diff
@@ -188,4 +201,53 @@ public Counter createForProjectView(Component projectView) {
       return null;
     }
   }
+
+  private static class LinesFormula implements Formula<LinesCounter> {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SizeMeasuresStep.java]: isn't this dead code? I don't see where `LinesFormula` is used

File: sonar-core/src/main/java/org/sonar/core/metric/ScannerMetrics.java [hunk 7]
```diff
@@ -74,7 +74,8 @@
 @ScannerSide
 public class ScannerMetrics {
 
-  private static final Set<Metric> ALLOWED_CORE_METRICS = ImmutableSet.<Metric>of(
+  private static final Set<Metric> ALLOWED_CORE_METRICS = ImmutableSet.of(
+    // Lines metric should be removed from this list when we are sure that no existing language plugins is sending measures on it
```
Comment (sns-seb) [inline sonar-core/src/main/java/org/sonar/core/metric/ScannerMetrics.java]: missing a `TODO` or `FIXME`

File: server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/component/ComponentRootBuilderTest.java (hunk 4 above)
Comment (henryju) [inline server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/component/ComponentRootBuilderTest.java]: File can be empty, so having 0 lines is possible.

Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/component/ComponentRootBuilderTest.java]: Yes it is ! I've already checked with scanner team, and even empty file have a number of lines to 1

File: sonar-scanner-protocol/src/main/protobuf/scanner_report.proto (hunk 5 above)
Comment (henryju) [inline sonar-scanner-protocol/src/main/protobuf/scanner_report.proto]: @dbmeneses do you confirm we can have files with 0 lines (empty files)

Comment (julienlancelot) [inline sonar-scanner-protocol/src/main/protobuf/scanner_report.proto]: Already checked, that's why I've added it

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SizeMeasuresStep.java (hunk 6 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SizeMeasuresStep.java]: Indeed, I'll remove it

File: sonar-scanner-protocol/src/main/protobuf/scanner_report.proto (hunk 5 above)
Comment (henryju) [inline sonar-scanner-protocol/src/main/protobuf/scanner_report.proto]: OK, sorry, I was pretty sure of the opposite.

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/FileAttributes.java (hunk 1 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/FileAttributes.java]: I can update the message, but I don't understand why should I put the check in ComponentRootBuilder

File: server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/component/ComponentRootBuilderTest.java (hunk 4 above)
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/component/ComponentRootBuilderTest.java]: @henryju Duarte has confirmed me that it was not possible

File: sonar-core/src/main/java/org/sonar/core/metric/ScannerMetrics.java [hunk 8]
```diff
@@ -74,7 +74,8 @@
 @ScannerSide
 public class ScannerMetrics {
 
-  private static final Set<Metric> ALLOWED_CORE_METRICS = ImmutableSet.<Metric>of(
+  private static final Set<Metric> ALLOWED_CORE_METRICS = ImmutableSet.of(
+    // TODO Lines metric should be removed from this list when we are sure that no existing language plugins is sending measures on it
```
Comment (SonarTech) [inline sonar-core/src/main/java/org/sonar/core/metric/ScannerMetrics.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'

**Comment by julienlancelot** [comment]
Commits of this PR has been merged in https://github.com/SonarSource/sonarqube/pull/1576

**Pull request closed without merging**